File: dashboard/utils/formatting.py
"""Pure data-shaping helpers used by page callbacks. Kept free of any Dash /
Flask imports so they're unit-testable without a running app or browser."""
from datetime import datetime
from typing import Optional

import plotly.graph_objects as go

from dashboard.theme import DASHBOARD_TEMPLATE_NAME, risk_color

CORE_VITAL_CHANNELS = ["HR", "O2Sat", "Temp", "SBP", "MAP", "Resp"]


def format_timestamp(value: Optional[str]) -> str:
    if not value:
        return "--"
    if isinstance(value, datetime):
        dt = value
    else:
        dt = datetime.fromisoformat(str(value).replace("Z", "+00:00"))
    return dt.strftime("%Y-%m-%d %H:%M:%S UTC")


def summarize_alerts(alerts: list[dict]) -> dict[str, int]:
    counts = {"Low": 0, "Medium": 0, "High": 0, "Critical": 0}
    for alert in alerts:
        level = alert.get("risk_level")
        if level in counts:
            counts[level] += 1
    return counts


def summarize_patients(patients: list[dict]) -> dict[str, int]:
    counts = {"active": 0, "discharged": 0, "deceased": 0}
    for patient in patients:
        status = patient.get("current_status")
        if status in counts:
            counts[status] += 1
    return counts


def build_vitals_figure(history: list[dict], channels: Optional[list[str]] = None) -> go.Figure:
    channels = channels or CORE_VITAL_CHANNELS
    fig = go.Figure()
    if not history:
        fig.update_layout(template=DASHBOARD_TEMPLATE_NAME, annotations=[_no_data_annotation()])
        return fig

    ordered = sorted(history, key=lambda v: v["timestamp"])
    timestamps = [v["timestamp"] for v in ordered]
    for channel in channels:
        values = [v.get("channels", {}).get(channel) for v in ordered]
        if all(value is None for value in values):
            continue
        fig.add_trace(go.Scatter(
            x=timestamps, y=values, mode="lines+markers", name=channel, connectgaps=True,
            line=dict(width=2.5, shape="spline", smoothing=0.3), marker=dict(size=5),
        ))
    fig.update_layout(
        template=DASHBOARD_TEMPLATE_NAME, legend_title_text="Channel", xaxis_title="Time", yaxis_title="Value",
        legend=dict(orientation="h", yanchor="bottom", y=1.02, xanchor="left", x=0),
    )
    return fig


def build_prediction_trend_figure(predictions: list[dict]) -> go.Figure:
    fig = go.Figure()
    if not predictions:
        fig.update_layout(template=DASHBOARD_TEMPLATE_NAME, annotations=[_no_data_annotation()])
        return fig

    ordered = sorted(predictions, key=lambda p: p["created_at"])
    timestamps = [p["created_at"] for p in ordered]
    probabilities = [p["sepsis_probability"] for p in ordered]
    marker_colors = [risk_color(p["risk_level"]) for p in ordered]

    fig.add_trace(go.Scatter(
        x=timestamps, y=probabilities, mode="lines+markers", name="Sepsis probability",
        line=dict(color="#38bdf8", width=3, shape="spline", smoothing=0.3),
        marker=dict(color=marker_colors, size=11, line=dict(width=2, color="#10141f")),
        fill="tozeroy", fillcolor="rgba(56, 189, 248, 0.08)",
    ))
    fig.add_hline(y=0.5, line_dash="dash", line_color=risk_color("Critical"), annotation_text="Critical threshold",
                  annotation_font_color=risk_color("Critical"))
    fig.update_layout(
        template=DASHBOARD_TEMPLATE_NAME, xaxis_title="Time", yaxis_title="Sepsis probability",
        yaxis_range=[0, 1], showlegend=False,
    )
    return fig


def build_shap_figure(shap_values: dict[str, float], top_n: int = 10) -> go.Figure:
    fig = go.Figure()
    if not shap_values:
        fig.update_layout(template=DASHBOARD_TEMPLATE_NAME, annotations=[_no_data_annotation()])
        return fig

    ranked = sorted(shap_values.items(), key=lambda kv: abs(kv[1]), reverse=True)[:top_n]
    ranked.reverse()  # largest contribution at top of a horizontal bar chart
    features = [k for k, _ in ranked]
    contributions = [v for _, v in ranked]
    colors = [risk_color("Critical") if v > 0 else risk_color("Low") for v in contributions]

    fig.add_trace(go.Bar(
        x=contributions, y=features, orientation="h", marker_color=colors,
        marker_line_width=0, text=[f"{v:+.3f}" for v in contributions], textposition="outside",
    ))
    fig.update_layout(
        template=DASHBOARD_TEMPLATE_NAME, xaxis_title="SHAP contribution (+ increases risk)", yaxis_title="Channel",
        bargap=0.35, uniformtext_minsize=10,
    )
    return fig


def _no_data_annotation() -> dict:
    return dict(text="No data yet", showarrow=False, xref="paper", yref="paper", x=0.5, y=0.5, font=dict(size=16))

File: frontend/src/utils/formatting.js
// Ported from dashboard/utils/formatting.py. Figure builders return Plotly
// figure objects ({ data, layout }) consumed by react-plotly.js <Plot>.
import { plotlyLayoutBase, riskColor } from "./theme";

export const CORE_VITAL_CHANNELS = ["HR", "O2Sat", "Temp", "SBP", "MAP", "Resp"];

export function formatTimestamp(value) {
  if (!value) return "--";
  const dt = new Date(value);
  if (Number.isNaN(dt.getTime())) return String(value);
  return dt.toISOString().replace("T", " ").slice(0, 19) + " UTC";
}

export function summarizeAlerts(alerts) {
  const counts = { Low: 0, Medium: 0, High: 0, Critical: 0 };
  for (const alert of alerts) {
    if (alert.risk_level in counts) counts[alert.risk_level] += 1;
  }
  return counts;
}

export function summarizePatients(patients) {
  const counts = { active: 0, discharged: 0, deceased: 0 };
  for (const patient of patients) {
    if (patient.current_status in counts) counts[patient.current_status] += 1;
  }
  return counts;
}

function noDataAnnotation() {
  return [{ text: "No data yet", showarrow: false, xref: "paper", yref: "paper", x: 0.5, y: 0.5, font: { size: 16 } }];
}

export function buildVitalsFigure(history, channels = CORE_VITAL_CHANNELS) {
  if (!history || history.length === 0) {
    return { data: [], layout: { ...plotlyLayoutBase, annotations: noDataAnnotation() } };
  }
  const ordered = [...history].sort((a, b) => new Date(a.timestamp) - new Date(b.timestamp));
  const timestamps = ordered.map((v) => v.timestamp);

  const data = [];
  for (const channel of channels) {
    const values = ordered.map((v) => v.channels?.[channel] ?? null);
    if (values.every((v) => v === null)) continue;
    data.push({
      x: timestamps,
      y: values,
      mode: "lines+markers",
      name: channel,
      connectgaps: true,
      line: { width: 2.5, shape: "spline", smoothing: 0.3 },
      marker: { size: 5 },
    });
  }
  return {
    data,
    layout: {
      ...plotlyLayoutBase,
      legend: { ...plotlyLayoutBase.legend, orientation: "h", yanchor: "bottom", y: 1.02, xanchor: "left", x: 0, title: { text: "Channel" } },
      xaxis: { ...plotlyLayoutBase.xaxis, title: { text: "Time" } },
      yaxis: { ...plotlyLayoutBase.yaxis, title: { text: "Value" } },
    },
  };
}

export function buildPredictionTrendFigure(predictions) {
  if (!predictions || predictions.length === 0) {
    return { data: [], layout: { ...plotlyLayoutBase, annotations: noDataAnnotation() } };
  }
  const ordered = [...predictions].sort((a, b) => new Date(a.created_at) - new Date(b.created_at));
  const timestamps = ordered.map((p) => p.created_at);
  const probabilities = ordered.map((p) => p.sepsis_probability);
  const markerColors = ordered.map((p) => riskColor(p.risk_level));

  return {
    data: [
      {
        x: timestamps,
        y: probabilities,
        mode: "lines+markers",
        name: "Sepsis probability",
        line: { color: "#38bdf8", width: 3, shape: "spline", smoothing: 0.3 },
        marker: { color: markerColors, size: 11, line: { width: 2, color: "#10141f" } },
        fill: "tozeroy",
        fillcolor: "rgba(56, 189, 248, 0.08)",
      },
    ],
    layout: {
      ...plotlyLayoutBase,
      xaxis: { ...plotlyLayoutBase.xaxis, title: { text: "Time" } },
      yaxis: { ...plotlyLayoutBase.yaxis, title: { text: "Sepsis probability" }, range: [0, 1] },
      showlegend: false,
      shapes: [
        {
          type: "line", x0: 0, x1: 1, xref: "paper", y0: 0.5, y1: 0.5, yref: "y",
          line: { color: riskColor("Critical"), dash: "dash" },
        },
      ],
      annotations: [
        { x: 1, y: 0.5, xref: "paper", yref: "y", text: "Critical threshold", showarrow: false,
          font: { color: riskColor("Critical") }, xanchor: "right", yanchor: "bottom" },
      ],
    },
  };
}

export function buildShapFigure(shapValues, topN = 10) {
  if (!shapValues || Object.keys(shapValues).length === 0) {
    return { data: [], layout: { ...plotlyLayoutBase, annotations: noDataAnnotation() } };
  }
  const ranked = Object.entries(shapValues)
    .sort((a, b) => Math.abs(b[1]) - Math.abs(a[1]))
    .slice(0, topN)
    .reverse();
  const features = ranked.map(([k]) => k);
  const contributions = ranked.map(([, v]) => v);
  const colors = contributions.map((v) => (v > 0 ? riskColor("Critical") : riskColor("Low")));

  return {
    data: [
      {
        x: contributions,
        y: features,
        type: "bar",
        orientation: "h",
        marker: { color: colors, line: { width: 0 } },
        text: contributions.map((v) => (v >= 0 ? `+${v.toFixed(3)}` : v.toFixed(3))),
        textposition: "outside",
      },
    ],
    layout: {
      ...plotlyLayoutBase,
      xaxis: { ...plotlyLayoutBase.xaxis, title: { text: "SHAP contribution (+ increases risk)" } },
      yaxis: { ...plotlyLayoutBase.yaxis, title: { text: "Channel" } },
      bargap: 0.35,
      uniformtext: { minsize: 10 },
    },
  };
}
